DOCSP-34184: Atlas Search indexes (#99)
mongodb/docs-rust: source/fundamentals/indexes/atlas-search-indexes.txt

.. _rust-atlas-search-indexes:

====================
Atlas Search Indexes
====================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, full text, atlas deployment

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to create and manage **Atlas Search indexes**. These indexes

> allow you to use MongoDB's Atlas Search feature, which performs fast, full-text searches

Author: I changed "which performs" to "to perform" and I think it resolves the personification issue, lmk what you think

on data stored in an Atlas cluster. The search indexes specify the behavior of the search
and which fields to index.

This guide includes the following sections:

- :ref:`rust-create-search-index`
- :ref:`rust-create-search-indexes`
- :ref:`rust-list-search-index`
- :ref:`rust-update-search-index`
- :ref:`rust-drop-search-index`

.. note:: 

   The examples in this guide access the ``posts`` collection in the ``sample_training``
   database, which is one of the Atlas sample datasets. For instructions on importing
   the Atlas sample data, see the :atlas:`Load Sample Data </sample-data>` page.

.. _rust-create-search-index:

Create a Search Index
---------------------

You can create an Atlas Search index on a single field or on multiple fields. Specify

> which fields to index in one of the following ways:

Author: They don't need to be strings; there are a few data types that aren't supported, but most are

- Use static mappings to specify the target fields
- Use dynamic mappings to instruct Atlas Search to automatically index all

>   supported field types

Author: I think it's more clear now after some restructuring, but lmk what you think

Example
~~~~~~~

The following example creates an Atlas Search index on the ``posts`` collection.
The code instantiates a ``SearchIndexModel`` and uses the struct's builder methods
to set the index name and definition. Then, the code passes the ``SearchIndexModel``
instance to the ``create_search_index()`` method to create an index that dynamically
indexes collection fields:

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/indexes.rs
      :start-after: begin-atlas-create-one
      :end-before: end-atlas-create-one
      :language: rust
      :dedent:

   .. output::
      :language: console
      :visible: false

      Created Atlas Search index:
      "example_index"

.. _rust-create-search-indexes:

Create Multiple Search Indexes
------------------------------

You can create multiple Atlas Search indexes at once by calling the ``create_search_indexes()``
method on your collection. This method accepts the following parameters:

- Index models for each index, specified in a vector of ``SearchIndexModel`` instances
- Index options, specified in a ``CreateSearchIndexOptions`` instance

Example
~~~~~~~

> The following example creates two Atlas Search indexes named ``"dynamic_index"`` and
> ``"static_index"`` on the ``posts`` collection. The code specifies ``SearchIndexModel``

Reviewer: S: since these are not string values but index names, I don't think they need "" quotes. Applies throughout where index names are mentioned.
```suggestion
The following example creates two Atlas Search indexes, ``dynamic_index`` and
``static_index``, on the ``posts`` collection. The code specifies ``SearchIndexModel``
```

instances for each index, defining the index names and definitions by using the struct's
builder methods. Then, the code passes these models as a vector to the ``create_search_indexes()``
method and creates the indexes:

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/indexes.rs
      :start-after: begin-atlas-create-many
      :end-before: end-atlas-create-many
      :language: rust
      :dedent:

   .. output::
      :language: console
      :visible: false

      Created Atlas Search indexes:
      ["dynamic_index", "static_index"]

.. _rust-list-search-index:

List Search Indexes
-------------------

> You can access information about your collection's Atlas Search indexes
> by calling the ``list_search_indexes()`` method on your collection. This
> method accepts the following parameters:

Reviewer: ```suggestion
You can access information about a collection's existing Atlas Search indexes
by calling the ``list_search_indexes()`` method on the collection. This
method accepts the following parameters:
```

- Name of the index to retrieve information about
- Aggregation options, specified in an ``AggregateOptions`` instance
- Index options, specified in a ``ListSearchIndexOptions`` instance

Example
~~~~~~~

The following example accesses information about the Atlas Search indexes on
the ``posts`` collection. The code calls the ``list_search_indexes()`` method
and passes a value of ``None`` for each parameter, which instructs the driver

> to return information about all Atlas Search indexes without configuring options.

Reviewer: ```suggestion
to return information about all Atlas Search indexes with default options.
```

> Then, the code iterates through the indexes and prints their information:

Reviewer: ```suggestion
Then, the code outputs the names of the Search indexes:
```

Author: Fixed with slightly different wording, since the example outputs all index information (not just the names)

.. literalinclude:: /includes/fundamentals/code-snippets/indexes.rs
   :start-after: begin-atlas-list
   :end-before: end-atlas-list
   :language: rust

>    :dedent:

Reviewer: I: should there be output here?

.. _rust-atlas-search-indexes:

====================
Atlas Search Indexes
====================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, full text, atlas deployment

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to create and manage **Atlas Search indexes**. These indexes
allow you to use MongoDB's Atlas Search feature to perform fast, full-text searches

> on data stored in an Atlas cluster. The Atlas Search index configures the behavior of Atlas

Reviewer: ```suggestion
on data stored in an Atlas cluster. An Atlas Search index configures the behavior of Atlas
```

Search by specifying which fields to index, how these fields are indexed, and other optional
settings.

This guide explains how to perform the following actions to manage your Atlas Search indexes:

- :ref:`rust-create-model`
- :ref:`rust-create-search-index`
- :ref:`rust-create-search-indexes`
- :ref:`rust-list-search-index`
- :ref:`rust-update-search-index`
- :ref:`rust-drop-search-index`

.. note:: 

   The examples in this guide access the ``posts`` collection in the ``sample_training``
   database, which is one of the Atlas sample datasets. For instructions on importing
   the Atlas sample data, see :atlas:`Load Sample Data </sample-data>` in the Atlas
   documentation.

.. _rust-create-model:

Create a Search Index Model
---------------------------

> Before creating an Atlas Search index, you must build a ``SearchIndexModel`` instance
> that sets your index specifications. To begin building a ``SearchIndexModel`` instance,
> call the ``SearchIndexModel::builder()`` method.

Reviewer: S: I think starting this paragraph with "Before creating a search index..." is confusing because building a `SearchIndexModel` is part of the process of creating the search index.
Perhaps this could clarify:

```suggestion
To create an Atlas Search index, you must first must build a ``SearchIndexModel`` instance
that sets your index specifications. To begin building a ``SearchIndexModel`` instance,
call the ``SearchIndexModel::builder()`` method.
```

> .. note:: Instantiating Options

Reviewer: S: since this is not an options object, rename the title
```suggestion
.. note:: Instantiating Models
```

   
   The {+driver-short+} implements the Builder design pattern for the
   creation of many different types, including ``SearchIndexModel``. You
   can use the ``builder()`` method to construct an instance of each type
   by chaining option builder methods.

The {+driver-short+} provides the following ``SearchIndexModel`` builder methods:

- ``definition()``, which accepts a BSON document parameter and sets your index definition
- ``name()``, which accepts a string parameter and sets your index name 

The BSON document that you pass to the ``definition()`` builder method must include
the ``mappings`` field. To automatically index all supported fields in your collection,
enable dynamic mappings by setting the ``mappings.dynamic`` nested field to ``true``. To

> index only certain specified fields, enable static mappings by setting the ``mappings.dynamic``

Reviewer: S: redundant
```suggestion
index only specified fields, enable static mappings by setting the ``mappings.dynamic``
```

> nested field to ``false`` and including a list of target fields.

Reviewer: ```suggestion
nested field to ``false`` and including a list of fields you want to index.
```

.. tip:: Atlas Search Field Mappings

   To learn more about Atlas Search field mappings, see :atlas:`Define Field Mapping
   </atlas-search/define-field-mappings/>` in the Atlas documentation.

Example
~~~~~~~

> The following example creates instructions for a index named ``example_index`` in a

Reviewer: S: "Creates instructions" in this context sounds incomplete
```suggestion
The following example creates specifications for a index named ``example_index`` in a
```

> ``SearchIndexModel`` instance. The code instructs Atlas Search to index only the ``body``
> and ``date`` fields by enabling static mappings:

Reviewer: ```suggestion
``SearchIndexModel`` instance. The code sets a static mapping to index only the ``body``
and ``date`` fields:
```

.. literalinclude:: /includes/fundamentals/code-snippets/indexes.rs
   :start-after: begin-atlas-model
   :end-before: end-atlas-model
   :language: rust
   :dedent:

.. _rust-create-search-index:

Create a Search Index
---------------------

> You can create an Atlas Search index by calling the ``create_search_index()`` method on
> your collection. This method accepts the following parameters:

Reviewer: ```suggestion
You can create an Atlas Search index on a collection by calling the ``create_search_index()`` method on
a ``Collection`` instance. This method accepts the following parameters:
```

Reviewer: change applies to all sections

- Index model, specified in a ``SearchIndexModel`` instance
- Index options, specified in a ``CreateSearchIndexOptions`` instance

Example
~~~~~~~

The following example creates an Atlas Search index on the ``posts`` collection.

> The code instantiates a ``SearchIndexModel`` and uses the struct's builder methods
> to set the index name and definition. Then, the code passes the ``SearchIndexModel``

Reviewer: S: remove unnecessary details
```suggestion
The code creates a ``SearchIndexModel`` that sets dynamic mapping and the index name.
Then, the code passes the ``SearchIndexModel``
```

> instance to the ``create_search_index()`` method to dynamically index collection fields:

Reviewer: ```suggestion
instance to the ``create_search_index()`` method to create the Search index:
```

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/indexes.rs
      :start-after: begin-atlas-create-one
      :end-before: end-atlas-create-one
      :language: rust
      :dedent:

   .. output::
      :language: console
      :visible: false

      Created Atlas Search index:
      "example_index"

.. _rust-create-search-indexes:

Create Multiple Search Indexes
------------------------------

You can create multiple Atlas Search indexes at once by calling the ``create_search_indexes()``
method on your collection. This method accepts the following parameters:

- List of index models, specified as a vector of ``SearchIndexModel`` instances
- Index options, specified in a ``CreateSearchIndexOptions`` instance

Example
~~~~~~~

The following example creates two Atlas Search indexes named ``dynamic_index`` and

> ``static_index`` on the ``posts`` collection. The code specifies ``SearchIndexModel``
> instances for each index, defining the index names and definitions by using the struct's
> builder methods. Then, the code passes these models as a vector to the ``create_search_indexes()``

Reviewer: ```suggestion
``static_index`` on the ``posts`` collection. The code creates ``SearchIndexModel``
instances for each index that specify the index names and definitions. Then, the code passes these models
as a vector to the ``create_search_indexes()``
```

mappings instead of static mappings. Then, the code calls the ``update_search_index()``
method to update the index:

.. literalinclude:: /includes/fundamentals/code-snippets/indexes.rs
   :start-after: begin-atlas-update
   :end-before: end-atlas-update
   :language: rust
   :dedent:

.. _rust-drop-search-index:

Delete a Search Index
---------------------

You can delete an Atlas Search index by calling the ``delete_search_index()``
method on a ``Collection`` instance. This method accepts the following parameters:

- Name of the index to delete
- Index options, specified in a ``DropSearchIndexOptions`` instance

Example
~~~~~~~

The following example deletes the Atlas Search index named ``example_index``
created in the :ref:`rust-create-search-index` section of this page. The code
passes the index name to the ``delete_search_index()`` method to delete the index:

.. literalinclude:: /includes/fundamentals/code-snippets/indexes.rs
   :start-after: begin-atlas-drop
   :end-before: end-atlas-drop
   :language: rust
   :dedent:
   
Additional Information
----------------------

To learn more about Atlas Search indexes, see the :atlas:`Atlas Search documentation

> </atlas-search/>`.

Reviewer: S: I think we should link this in the overview of this page, as users might get to this page and want to read about the feature before reading the driver documentation. This section could link back out to the indexes page, and here are a few more pages we could link here:
- https://www.mongodb.com/docs/atlas/atlas-search/best-practices
- https://www.mongodb.com/docs/atlas/atlas-search/performance/
- https://www.mongodb.com/docs/atlas/atlas-search/limitations/
- https://www.mongodb.com/docs/atlas/atlas-search/faq/

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about the methods and types mentioned in this
guide, see the following API documentation:

- `create_search_index() <{+api+}/struct.Collection.html#method.create_search_index>`__
- `SearchIndexModel <{+api+}/struct.SearchIndexModel.html>`__
- `CreateSearchIndexOptions <{+api+}/options/struct.CreateSearchIndexOptions.html>`__
- `create_search_indexes() <{+api+}/struct.Collection.html#method.create_search_indexes>`__
- `list_search_indexes() <{+api+}/struct.Collection.html#method.list_search_indexes>`__
- `AggregateOptions <{+api+}/options/struct.AggregateOptions.html>`__
- `ListSearchIndexOptions <{+api+}/options/struct.ListSearchIndexOptions.html>`__
- `update_search_index() <{+api+}/struct.Collection.html#method.create_search_index>`__
- `UpdateSearchIndexOptions <{+api+}/options/struct.UpdateSearchIndexOptions.html>`__
- `drop_search_index() <{+api+}/struct.Collection.html#method.create_search_index>`__

> - `DropSearchIndexOptions <{+api+}/options/struct.DropSearchIndexOptions.html>`__

Reviewer: S: Leaving this up to you, but since this list is quite long, could be easier to traverse links if you inlined them, as we do on the [DB/collection guide](https://preview-mongodbnorareidy.gatsbyjs.io/rust/DOCSP-34184-atlas-search-index/fundamentals/database-collection/) (for the same reason)
